File: src/main/java/udf/LatLng2Area.java
package udf;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import org.apache.hadoop.hive.ql.exec.UDF;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

import java.io.IOException;

public class LatLng2Area extends UDF {

    public String evaluate(String lat, String lng) {

        // 传入参数
        String json = null;
        try {
            json = getJson(lat, lng);
        } catch (IOException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        // 传入参数
        String res = getResult(json);
        // 返回结果
        return res;
    }

    public String getJson(String lat, String lng) throws IOException, InterruptedException {
        // lat  维度：0-90，lng:经度：0-180
        String lng_lat = lng + "," + lat;

        // 获取 url 腾讯 api 使用的是纬经度 （维度，经度）
//        String urlStr = "https://apis.map.qq.com/ws/geocoder/v1/?location=" + lat_lng + "&key=IWABZ-PM362-K4XU4-CAMZY-4Y5CK-UJBNI" + "&get_poi=1";

     // 高德地图 api 使用的是经纬度 （经度，维度）
        String urlStr = "https://restapi.amap.com/v3/geocode/regeo?location=" + lng_lat + "&key=a19f03acd91d1b995fe5dbb9cf069ccd" + "&radius=1000&extensions=all";

//https://restapi.amap.com/v3/geocode/regeo?location=116.310003,39.991957&key=a19f03acd91d1b995fe5dbb9cf069ccd&radius=1000&extensions=all

        // 获取http，返回的是json
        String res = getUrlResult(urlStr);

        return res;
    }

    public String getUrlResult(String urlStr) throws InterruptedException, IOException {
        // 获取 客户端
        CloseableHttpClient cli = HttpClients.createDefault();
        HttpGet httpGet = new HttpGet(urlStr);

        // 发送请求，间隔1秒
        Thread.sleep(1000);


        CloseableHttpResponse reponse = null;
        try {
            reponse = cli.execute(httpGet);
        } catch (IOException e) {
            e.printStackTrace();
        }
        // 格式化结果

        return EntityUtils.toString(reponse.getEntity(), "UTF-8");


    }



    public String getResult(String json) {

        // 解析 json
        JSONObject jsonObj = JSON.parseObject(json);

        // 判断状态
        int status = jsonObj.getIntValue("status");
        if (1 != status) {
            return "";
        }

        // 判断状态
        JSONObject resultArr = jsonObj.getJSONObject("regeocode");
        if (resultArr == null) {
            return "";
        }

        // 判断状态
        JSONObject addre = resultArr.getJSONObject("addressComponent");
        if (addre == null) {
            return "";
        }

        String province = addre.getString("province");
        String city = addre.getString("city");
        String district = addre.getString("district");
        String township = addre.getString("township");

//        JSONObject loca = resultArr.getJSONObject("location");
//        String lat = loca.getString("lat");
//        String lng = loca.getString("lng");

        String res =   province + "," + city + "," + district + "," + township;
        return res;
    }


//    public static void main(String[] args) throws IOException, InterruptedException {
//        System.out.println(new LatLng2Area().evaluate("31.4141", "118.233131"));
//    }

}

File: src/main/scala/com/black_unique/udf/Lat_lng_2_area.scala
package com.black_unique.udf

import com.alibaba.fastjson.{JSON, JSONObject}
import org.apache.commons.lang.StringUtils
import org.apache.hadoop.hive.ql.exec.UDF
import org.apache.http.client.methods.HttpGet
import org.apache.http.impl.client.HttpClients
import org.apache.http.util.EntityUtils

/**
  * @Auther: tjk
  * @Date: 2020-07-07 22:08
  * @Description:
  */

object Lat_lng_2_area extends UDF {

  def evaluate(lat: String, lng: String): String = {
    //判断传入的数据是否合法
    if (StringUtils.isEmpty(lat) || StringUtils.isEmpty(lng)) throw new RuntimeException("参数有误")
    // 传入参数
    val json = getJson(lat, lng)
    // 传入参数
    val res = getResult(json)
    // 返回结果
    res
  }

  /**
    * 传入url,返回字符串
    *
    * @param urlStr
    * @return
    */
  def getUrlResult(urlStr: String) = {

    // 获取 客户端
    val client = HttpClients.createDefault()
    val httpGet = new HttpGet(urlStr)

    // 发送请求，间隔1秒
    Thread.sleep(1000)

    val reponse = client execute httpGet
    // 格式化结果
    EntityUtils.toString(reponse.getEntity, "UTF-8")

  }

  /**
    * 传入经纬度，返回json传
    *
    * @param lat
    * @param lng
    * @return
    */
  def getJson(lat: String, lng: String): String = {

    // lat  维度：0-90，lng:经度：0-180
    val lat_lng = lat + "," + lng

    // 获取 url
    val urlStr = "https://apis.map.qq.com/ws/geocoder/v1/?location=" + lat_lng + "&key=IWABZ-PM362-K4XU4-CAMZY-4Y5CK-UJBNI" + "&get_poi=1"

    // 获取http，返回的是json
    val res = getUrlResult(urlStr)

    // 返回值:一行json数据
    res
  }

  /**
    * 传入json串，返回结果
    *
    * @param json
    * @return
    */
  def getResult(json: String): String = {
    // 解析 json
    val jsonObj: JSONObject = JSON.parseObject(json)

    // 判断状态
    val status = jsonObj.getIntValue("status")
    if (0 != status) {
      return ""
    }

    // 判断状态
    val resultArr = jsonObj.getJSONObject("result")
    if (resultArr == null) {
      return ""
    }

    // 判断状态
    val addre = resultArr.getJSONObject("address_component")
    if (addre == null) {
      return ""
    }

    val province = addre.getString("province")

    val city = addre.getString("city")
    val district = addre.getString("district")
    val street = addre.getString("street")

    val loca = resultArr.getJSONObject("location")
    val lat = loca.getString("lat")
    val lng = loca.getString("lng")

    val res = lat + "," + lng + "," + province + "," + city + "," + district + "," + street
    res

  }

  def main(args: Array[String]): Unit = {
    println(Lat_lng_2_area.evaluate("40.64178", "109.83859"))
  }
}